File: src/services/auth.service.ts
import { getAuth, signInWithEmailAndPassword, createUserWithEmailAndPassword, signOut } from 'firebase/auth';
import { UserRepository } from '../repositories/user.repository';
import { User } from '../models/user';

export class AuthService {
  private auth = getAuth();
  private userRepository = new UserRepository();

  async signUp(email: string, password: string, tenantId: string): Promise<void> {
    const userCredential = await createUserWithEmailAndPassword(this.auth, email, password);
    const user: User = {
      id: userCredential.user.uid,
      email,
      roles: ['STUDENT'],
      tenantId,
      provider: 'email'
    };
    await this.userRepository.createUser(user);
  }

  async signOut(): Promise<void> {
    await signOut(this.auth);
  }
}

File: mobile_app/lib/services/auth_service.dart
import 'package:flutter/foundation.dart';
import 'package:firebase_auth/firebase_auth.dart';

class AuthService extends ChangeNotifier {
  final FirebaseAuth _auth = FirebaseAuth.instance;

  User? get currentUser => _auth.currentUser;

  Stream<User?> authStateChanges() => _auth.authStateChanges();

  Future<UserCredential?> signUp({
    required String email,
    required String password,
  }) async {
    try {
      final credential = await _auth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );
      notifyListeners();
      return credential;
    } on FirebaseAuthException catch (e) {
      if (e.code == 'weak-password') {
        throw 'A senha é muito fraca.';
      } else if (e.code == 'email-already-in-use') {
        throw 'Este email já está registado.';
      } else {
        throw 'Erro ao registar: ${e.message}';
      }
    }
  }

  Future<UserCredential?> signIn({
    required String email,
    required String password,
  }) async {
    try {
      final credential = await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
      notifyListeners();
      return credential;
    } on FirebaseAuthException catch (e) {
      if (e.code == 'user-not-found') {
        throw 'Utilizador não encontrado.';
      } else if (e.code == 'wrong-password') {
        throw 'Senha incorreta.';
      } else {
        throw 'Erro ao iniciar sessão: ${e.message}';
      }
    }
  }

  Future<void> signOut() async {
    await _auth.signOut();
    notifyListeners();
  }

  Future<void> resetPassword(String email) async {
    await _auth.sendPasswordResetEmail(email: email);
  }
}
